### PR TITLE
Bash completion

### PR DESCRIPTION
**Pull request by aude**
Hi, Axel!

I like doing [Hacktoberfest](https://hacktoberfest.digitalocean.com/) and thought it would be a good opportunity to work on Bash completion for `btrbk` (#298).

Here's an MVP, where options completion is fully implemented. Commands completion is not implemented.

If you accept, it would be awesome if you added the `hacktoberfest-accepted` label to this PR. ([reason](https://hacktoberfest.digitalocean.com/hacktoberfest-update))

### CODE REVIEW

**Comment by digint** [comment]
Thanks for contributing! (and sorry for the late reply, I'm super-busy with other things at the moment...)

I somehow miss the "commands" (run, snapshot, ...) in the completions. Note that I have no clue how the magic of `compgen` works, maybe I'm missing something here...

Also, a completion for the new `lsbtr` command would be nice!

**Comment by aude** [comment]
You're right, I haven't implemented completion for the "commands", was planning on doing that in a future pull request. Was thinking it was better to get something shipped and then get motivation to improve it, does that strategy work for you?

Agree, completion for `lsbtr` would be nice. Would probably be a breeze after completion for `btrbk` is done.

**Comment by digint** [comment]
I'd rather wait with merging until the commands are also available in the completion, I don't like to have half-finished stuff in master. Will gladly merge it when finished, thanks for your efforts!

**Comment by aude** [comment]
Alright, thanks for being clear!

I'll work on completion for commands today.

**Comment by aude** [comment]
Done :heavy_check_mark:

**Comment by digint** [comment]
had a quick look at it, works for me. it;s a bit odd that i have to specify the `--xxx` options at the beginning (this seems to be some completion magic?), while you can actually put them everywhere (no, not like git, *everywhere*  in the command arguments ;-). will have a closer look later, can't push right now anyways...

**Comment by aude** [comment]
Oh! I actually had to make quite the effort to make that work, doh :laughing:  I did it because of:
```
usage: btrbk [options] <command> [filter...]
```
So I interpreted that as "`[options]` are intended to be specified before `<command>`".

But now I know what's intended, so I'll look into completing options anywhere :+1:

**Comment by aude** [comment]
Pushed a new version that completes options everywhere.

File: contrib/bash/completion.bash [hunk 1]
```diff
@@ -16,94 +16,76 @@ _btrbk_init_cmds()
...
+    if [[ ! -v 'cmds[0]' ]]; then
+      COMPREPLY=($(compgen -W 'run dryrun snapshot resume prune archive clean stats list usage origin diff ls' -- "$cur"))
+    fi
+  fi
 
   case "${cmds[0]}" in
     'archive')
       # <source>
       if [[ ! -v 'cmds[1]' ]]; then
+        echo 'a' 1>&2
```
Comment (digint) [inline contrib/bash/completion.bash:62]: is this some required completion magic, or just forgotten debug code?

**Comment by digint** [comment]
i tried to follow "best-practice" here:
```
usage: btrbk [options] <command> [filter...]
```

e.g. `ls --help` shows similar:
```
# ls --help
Usage: ls [OPTION]... [FILE]...
```

I think it's pretty clear, but does not hint you it also works vice-versa (I understand this as "best practice", as most command line tools support it).

see comments on code, works nicely for me (on gentoo linux)

File: contrib/bash/completion.bash (hunk 1 above)
Comment (aude) [reply contrib/bash/completion.bash:62]: It's just forgotten debug code, thanks, I'll remove it.

**Comment by aude** [comment]
That makes sense. I don't know how it could be written differently. And it's true, many commands support options everywhere (which is the most convenient anyway). So it was just me interpreting it the wrong way I think.

**Comment by digint** [comment]
merged to master, `hacktoberfest-accepted`

thanks again, now let's just hope i don't forget to keep this up-to-date on releases.
I think this will prove useful for many people. Will probably also need to do adaptions in:  https://github.com/digint/btrbk-packaging/tree/debian/debian, should be straight forward.

**Comment by aude** [comment]
Awesome! Thanks for reviewing. Just ping me at any point if you need me to explain something about the code at any point.

**Pull request merged**